File: packages/go/echozero/inbound_canonical.go
package echozero

import (
	"encoding/json"
	"math"
	"sort"
	"strings"
)

var structuredSigningKeys = []string{
	"action",
	"amount",
	"chain",
	"changes",
	"clientTimestamp",
	"confidence",
	"context",
	"currentPnlPct",
	"entryPrice",
	"entryZone",
	"eventType",
	"exitPrice",
	"exitReason",
	"expiresAt",
	"ideaId",
	"instrument",
	"leverageX",
	"metadata",
	"orderType",
	"pnlPct",
	"positionRef",
	"reasoning",
	"relatedTradeId",
	"riskMgmt",
	"sellSizePct",
	"side",
	"symbol",
	"tokenAddress",
	"tradeType",
	"triggers",
	"urgency",
	"version",
}

func canonicalizeValue(value any) any {
	if value == nil {
		return nil
	}
	switch typed := value.(type) {
	case string:
		return typed
	case bool:
		return typed
	case int:
		return float64(typed)
	case int32:
		return float64(typed)
	case int64:
		return float64(typed)
	case float32:
		if math.IsNaN(float64(typed)) || math.IsInf(float64(typed), 0) {
			return nil
		}
		return float64(typed)
	case float64:
		if math.IsNaN(typed) || math.IsInf(typed, 0) {
			return nil
		}
		return typed
	case json.Number:
		if i, err := typed.Int64(); err == nil {
			return float64(i)
		}
		if f, err := typed.Float64(); err == nil {
			if math.IsNaN(f) || math.IsInf(f, 0) {
				return nil
			}
			return f
		}
		return nil
	case []any:
		out := make([]any, 0, len(typed))
		for _, item := range typed {
			if child := canonicalizeValue(item); child != nil || item == nil {
				out = append(out, child)
			}
		}
		return out
	case map[string]any:
		out := make(map[string]any)
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			child := canonicalizeValue(typed[key])
			if child != nil || typed[key] == nil {
				out[key] = child
			}
		}
		return out
	default:
		return nil
	}
}

func InboundWebhookCanonicalJSON(body map[string]any) (string, error) {
	sorted := make(map[string]any)

	if text, ok := body["text"].(string); ok {
		sorted["text"] = text
	}
	if idk, ok := body["idempotencyKey"].(string); ok {
		trimmed := strings.TrimSpace(idk)
		if trimmed != "" {
			sorted["idempotencyKey"] = trimmed
		}
	}

	for _, key := range structuredSigningKeys {
		if value, ok := body[key]; ok {
			if child := canonicalizeValue(value); child != nil || value == nil {
				sorted[key] = child
			}
		}
	}

	keys := make([]string, 0, len(sorted))
	for key := range sorted {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	obj := make(map[string]any, len(keys))
	for _, key := range keys {
		obj[key] = sorted[key]
	}
	raw, err := json.Marshal(obj)
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

File: packages/python/src/echozero/inbound_canonical.py
"""Mirrors `agentInboundWebhook.signing.ts` in mcp-main."""

from __future__ import annotations

import json
from typing import Any, Mapping

STRUCTURED_SIGNING_KEYS = (
    "action",
    "amount",
    "chain",
    "changes",
    "clientTimestamp",
    "confidence",
    "context",
    "currentPnlPct",
    "entryPrice",
    "entryZone",
    "eventType",
    "exitPrice",
    "exitReason",
    "expiresAt",
    "ideaId",
    "instrument",
    "leverageX",
    "metadata",
    "orderType",
    "pnlPct",
    "positionRef",
    "reasoning",
    "relatedTradeId",
    "riskMgmt",
    "sellSizePct",
    "side",
    "symbol",
    "tokenAddress",
    "tradeType",
    "triggers",
    "urgency",
    "version",
)


def _canonicalize_value(value: Any) -> Any | None:
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value
    if isinstance(value, (int, float)):
        if isinstance(value, float) and (value != value or value in (float("inf"), float("-inf"))):
            return None
        return value
    if isinstance(value, list):
        return [_canonicalize_value(item) for item in value]
    if isinstance(value, dict):
        out: dict[str, Any] = {}
        for key in sorted(value):
            child = _canonicalize_value(value[key])
            if child is not None or value[key] is None:
                out[key] = child
        return out
    return None


def inbound_webhook_canonical_json(body: Mapping[str, Any]) -> str:
    sorted_body: dict[str, Any] = {}

    if "text" in body and body["text"] is not None:
        sorted_body["text"] = body["text"]

    idk = body.get("idempotencyKey")
    if isinstance(idk, str):
        trimmed = idk.strip()
        if trimmed:
            sorted_body["idempotencyKey"] = trimmed

    for key in STRUCTURED_SIGNING_KEYS:
        if key not in body:
            continue
        value = _canonicalize_value(body[key])
        if value is not None or body[key] is None:
            sorted_body[key] = value

    return json.dumps(
        {key: sorted_body[key] for key in sorted(sorted_body)},
        separators=(",", ":"),
        ensure_ascii=False,
    )


def rest_request_body_text(body: Any | None) -> str:
    if body is None:
        return ""
    if isinstance(body, str):
        return body
    return json.dumps(body, separators=(",", ":"), ensure_ascii=False)
